DO NOT MERGE -- Update the NEWS for 4.3 (#2503)
awesomeWM/awesome: docs/89-NEWS.md

# NEWS

<a name="v43"></a>
# Awesome window manager framework version 4.3 changes

<center> <img src="../images/AUTOGEN_wibox_logo_logo_and_name.svg" /> </center>

Awesome v4.3 is the third release of the 4.x API. It comes after a year of

> little fixes and improvements. Awesome v4.2 was very stable thanks to everybody

Reviewer: s/everybody/everybody's/

> little fixes and improvements. Awesome v4.2 was very stable thanks to everybody

Reviewer 2: ```suggestion
little fixes and improvements. Awesome v4.2 was very stable thanks to everybody's
```
(using Github's new feature)

effort to unit test everything. Given no major bug warranted a new release,
this one adds a few large features while preserving full compatibility with
existing user configurations.

## New features

>  * `gears.string` now has a `endswith` and `startswith` functions

Reviewer 2: ```suggestion
 * `gears.string` now has `endswith` and `startswith` functions
```

 * `luarocks` modules are now automatically available in Awesome

>  * A generic way to create or use widgets have been added

Reviewer 2: ```suggestion
 * A generic way to create or use widgets has been added
```

   (`wibox.widget.base.make_widget_from_value`)
 * It is now possible to connect to signals from all instances of a widget at once
 * The calendar widget now supports margins
 * The documentation has a new theme

>  * Wiboxes now have a `to_widget()` and `save_to_svg()` methods.

Reviewer 2: ```suggestion
 * Wiboxes now have a `to_widget()` and `save_to_svg()` method.
```

Reviewer 3: Keeping in line with the above gears.string grammar, would we want to keep `methods` and remove the `a` instead?

 * The client objects now have a `immobilized_horizontal` and
   `immobilized_vertical` property to know if they can currently be moved or
   resized (for example, it is set to false when maximized)
 * `gears.timer` objects now have a `call_now` method.

>  * The hoykey popup now supports `termite` keybindings

Reviewer 4: hoykey => hotkey?

 * The menubar now loads faster
 * The wiboxes now have an `input_passthrough` property to send mouse clicks
   to the object below.
 * The `taglist` and `tasklist` now support the declarative constructor syntax
 * There is now an `awesome.pixbuf_to_surface` to bypass some LGI limitations

>  * The notifications icon can now be resized and limited with

Reviewer 2: ```suggestion
 * The icon for notifications can now be resized and limited with
```

>    `notification\_icon_\size`

Reviewer 5: I don't think any backslashes are needed here

 * A `gears.sort` module has been added with graph resolution
 * `awesome-client` now runs code in a protected context
 * The <a href="./03-declarative-layout.md.html">widget documentation</a> has
   been extended to be more friendly to new users.
 * There is a new `beautiful.maximized_hide_border` theme option to hide the
   border for maximized clients.
 * The `client` `startup_id` field is now writable. This is useful when the

>    client native implementation is not present or too buggy to be used.

Reviewer 2: ```suggestion
   client's native implementation is not present or too buggy to be used.
```

 * The `awful.widget.prompt` now has a `with_shell` option to allow Bash/ZSH
   aliases, function and environment variables to be used in commands.
 * The `awful.titlebar`s now have a `fallback_name` when a client has no `name`
   property.
 * Clients now have a `motif_wm_hints` property to reflect some hints using the
   Motif X11 property extension. This is used by some modern toolkits including
   GTK.
 * Clients now have a `requests_no_titlebar` property to expose when a client
   has client side titlebars (also known as decorations and CSD)
 * The hotkey popup now has a `show_awesome_keys` option.
 * The `awful.widget.prompt` now has more of the `awful.prompt` constructor
   arguments.
 * It is now possible to set a list of layouts per tag instead of a single
   global one.
 * There is now a `awful.layout.get_tag_layout_index()` function to get the
   index of the current layout in the global layout list
   (`awful.layout..layouts`)

### Better DPI handling

The screen now has a read/write `dpi` property and `awful.screen.set_auto_dpi_enabled(true)`
can be used to automatically set the DPI for many Awesome elements. Please

> note that it is not backward compatible and breaks many widget. As AwesomeWM

Reviewer 2: ```suggestion
note that it is not backward compatible and breaks many widgets. Since AwesomeWM
```

Reviewer 3: `not backward compatible and will break many widgets.` ?

> always used pixels as the de-facto metric for sizes, enabling `anto_dpi` will

Reviewer 5: Typo in `auto_dpi`

> break most existing configs. However, for people who use such setup, it might be

Reviewer 2: ```suggestion
break most existing configs. However, for people who use such setups, it might be
```

> worth speding some time to fix their config.

Reviewer 2: ```suggestion
worth spending some time to fix their config.
```

### Extendable `awful.rules` providers and better `awful.spawn` functions

> There is two new functions called `awful.rules.add_rule_source` and

Reviewer 2: ```suggestion
There are two new functions called `awful.rules.add_rule_source` and
```

`awful.rules.remove_rule_source`. They allow to create a dependency graph
for where a rule comes from and which provider has the priority when setting it.

> Previously, there was the normal properties,

Reviewer 2: ```suggestion
Previously, there were the normal properties,
```

`awful.rules.high_priority_properties` and `awful.rules.delayed_properties`.

> This didn't scale and could not represent all corner cases. Those table still

Reviewer 2: ```suggestion
This didn't scale and could not represent all corner cases. Those tables still
```

exist and are still honored, but there is now a system that can handle the full
complexity of the property priority graph.

> Making use of this by default is `awful.spawn`. The reliability of the attaching

Reviewer 2: ```suggestion
`awful.spawn` is making use of this by default. The reliability of attaching
```

properties to the `spawn` calls has been improved. On top of this, 3 new
functions have been added

 * `awful.spawn.once`
 * `awful.spawn.single_instance`
 * `awful.spawn.raise_or_spawn`

They allow to specify a command can only have one running instance. This works

> across restart too, so all hacks to handle restarting Awesome are no longer

Reviewer 2: ```suggestion
across restarts too, so all hacks to handle restarting Awesome are no longer
```

> required.

Reviewer 2: ```suggestion
needed.
```

Note that the `client.startup_id` isn't supported by all applications and the
Linux specific workaround is recommended to improve the reliability of the
`awful.spawn` functions.

### A brand new keygrabber API

The `keygrabber` module API was rebuilt from scratch to make it more usable.

> The previous API was very low level, very close to how it actually work, but

Reviewer 2: ```suggestion
The previous API was very low level, very close to how it actually works, but
```

was disconnected from how keygrabbers are used in a window manager. Getting
anything done with the previous API required a lot of boilerplate code and had
many corner cases to handle. The new API has built-in support for the most
common use cases and is fully declarative.

### A new GTK color palette based theme

> A new theme has been added. It reads the GTK theme colors and use them in the

Reviewer 2: ```suggestion
A new theme has been added. It reads the GTK theme colors and uses them in the
```

> `wibar`, `menu` and `titlebar`. It helps create an uniform look and feel

Reviewer 2: ```suggestion
`wibar`, `menu` and `titlebar`. It helps creating an uniform look and feel
```

Reviewer 3: `uniform` uses `a` rather than `an`. `create a uniform` is the correct grammar.

> between the window manager and client applications with minimal efforts.

Reviewer 2: ```suggestion
between the window manager and client applications with minimal effort.
```

> ### Widgets improvments

Reviewer 6: typo: improvements

[![The separator widget](../images/AUTOGEN_wibox_layout_flex_spacing_widget.svg)](../classes/wibox.layout.flex.html#wibox.layout.flex.spacing_widget)

The `awful.widget.taglist` and `awful.widget.tasklist` now support creating
custom widgets for each elements:

[![The separator widget](../images/AUTOGEN_wibox_awidget_taglist_indexed.svg)](../classes/wibox.layout.flex.html#wibox.layout.flex.spacing_widget)

[![The separator widget](../images/AUTOGEN_wibox_awidget_tasklist_rounded.svg)](../classes/wibox.layout.flex.html#wibox.layout.flex.spacing_widget)

[![The separator widget](../images/AUTOGEN_wibox_awidget_tasklist_windows10.svg)](../classes/wibox.layout.flex.html#wibox.layout.flex.spacing_widget)

A new `popup` widget allows to bypass most of the boilerplate code and easily
display widgets on the screen:

[![The ratio strategies](../images/AUTOGEN_awful_popup_wiboxtypes.svg)](../classes/awful.popup.html)

The `awful.widget.layoutlist` allows to easily display and select the client
layout from a widget:

[![The layoutbox](../images/AUTOGEN_awful_widget_layoutlist_bar.svg)](../classes/awful.widget.layoutlist.html)

## Noteworthy fixes

>  * There is no longer an error when a tag defined by name in `awful.rules` it

Reviewer 2: ```suggestion
 * There is no longer an error when a tag defined by name in `awful.rules` is
```

   not found.

>  * The menubar is now generally more robust thanks to a variety of improvments

Reviewer 6: typo: improvements

 * Many dead links in the documentation have been fixed

>  * The `textclock` is not generally more robust handling formatting issues,

Reviewer 2: ```suggestion
 * The `textclock` is now generally more robust handling formatting issues,
```

   timezones and declarative constructors.
 * The last screen is never removed. Previously, buggy laptop suspend could

>    remove all the screens, causing all clients to go to the first tag or lost

Reviewer 2: ```suggestion
   remove all the screens, causing all clients to go to the first tag or getting lost
```

   completely.
 * The new default `rc.lua` now uses `request::activate` to set the focus. This

>    fixes many corner case such as unfocusable clients getting the focus

Reviewer 2: ```suggestion
   fixes many corner cases such as unfocusable clients getting the focus
```

 * Calling `awful.spawn` with a set of properties is now more reliable.

## Behavior changes

> * When trying to access a screen that doesn't exist. It now returns `nil` instead

Reviewer 2: ```suggestion
* When trying to access a screen that doesn't exist it now returns `nil` instead
```

>   asserting. So hypothetical code that relied on `pcall()` to detect the screen

Reviewer 2: ```suggestion
  of asserting. So hypothetical code that relied on `pcall()` to detect the screen
```

>   existance instead of using the proper API (`screen.count`) now need to be

Reviewer 2: ```suggestion
  existance instead of using the proper API (`screen.count`) now needs to be
```

 * There is a new `beautiful.maximized_hide_border` theme option to hide the
   border for maximized clients.
 * The `client` `startup_id` field is now writable. This is useful when the
   client native implementation is not present or too buggy to be used.
 * The `awful.widget.prompt` now has a `with_shell` option to allow Bash/ZSH
   aliases, function and environment variables to be used in commands.
 * The `awful.titlebar`s now have a `fallback_name` when a client has no `name`
   property.
 * Clients now have a `motif_wm_hints` property to reflect some hints using the
   Motif X11 property extension. This is used by some modern toolkits including
   GTK.
 * Clients now have a `requests_no_titlebar` property to expose when a client
   has client side titlebars (also known as decorations and CSD)
 * The hotkey popup now has a `show_awesome_keys` option.
 * The `awful.widget.prompt` now has more of the `awful.prompt` constructor
   arguments.
 * It is now possible to set a list of layouts per tag instead of a single
   global one.
 * There is now a `awful.layout.get_tag_layout_index()` function to get the
   index of the current layout in the global layout list

>    (`awful.layout..layouts`)

Reviewer 7: Is this `..` a typo? I'm not actually sure.

There is two new functions called `awful.rules.add_rule_source` and
`awful.rules.remove_rule_source`. They allow to create a dependency graph
for where a rule comes from and which provider has the priority when setting it.

Previously, there was the normal properties,
`awful.rules.high_priority_properties` and `awful.rules.delayed_properties`.
This didn't scale and could not represent all corner cases. Those table still
exist and are still honored, but there is now a system that can handle the full
complexity of the property priority graph.

Making use of this by default is `awful.spawn`. The reliability of the attaching
properties to the `spawn` calls has been improved. On top of this, 3 new
functions have been added

 * `awful.spawn.once`
 * `awful.spawn.single_instance`
 * `awful.spawn.raise_or_spawn`

They allow to specify a command can only have one running instance. This works
across restart too, so all hacks to handle restarting Awesome are no longer
required.

Note that the `client.startup_id` isn't supported by all applications and the
Linux specific workaround is recommended to improve the reliability of the

> `awful.spawn` functions.

Reviewer 7: Should this link to the workaround?
